### PR TITLE
Add pairing Agent

### PR DESCRIPTION
**Pull request by danrabbit**
Screenshot from 2019-04-07 12 07 54@2x

image

This needs to be tested with:
- [ ]  A device that needs confirmation before pairing
- [x] A device that provides a passcode which we are confirming before pressing pair on that device
- [ ] A device that is requesting we confirm a passkey before pressing pair on this device
- [x] A device that is requesting a pin be entered on that device

ToDo:

- [x] Dismiss dialog on success
- [ ] Show something on failure
- [x] Get the actual device name

### CODE REVIEW

**Comment by cassidyjames** [comment]
For keyboards that require a passcode, it's not a matter of confirming it; it requires the user to enter the code and press "Enter" on the keyboard they're pairing. Probably should mention that if the device type is keyboard?

File: src/PairDialog.vala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+public class PairDialog : Granite.MessageDialog {
+    public PairDialog (ObjectPath device) {
+        Object (
+            buttons: Gtk.ButtonsType.CANCEL,
+            image_icon: new ThemedIcon ("dialog-question"),
+            primary_text: _("Confirm Bluetooth Pairing"),
+            secondary_text: _("\"%s\" would like to pair with this device.").printf ("Helix-Sama")
```
Comment (cassidyjames) [inline src/PairDialog.vala]: Typographically-correct quotes:

```suggestion
            secondary_text: _("“%s” would like to pair with this device.").printf ("Helix-Sama")
```

File: src/PairDialog.vala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        var confirm_button = add_button (_("Pair"), Gtk.ResponseType.ACCEPT);
+        confirm_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+    }
+
+    public PairDialog.with_passkey (ObjectPath device, uint32 passkey) {
+        Object (
+            buttons: Gtk.ButtonsType.CANCEL,
+            image_icon: new ThemedIcon ("dialog-information"),
+            primary_text: _("Confirm Bluetooth Passkey"),
+            secondary_text: _("Make sure the code displayed on \"%s\" matches the one below.").printf ("Helix-Sama")
```
Comment (cassidyjames) [inline src/PairDialog.vala]: Typographically-correct quotes:

```suggestion
            secondary_text: _("Make sure the code displayed on “%s” matches the one below.").printf ("Helix-Sama")
```

File: src/PairDialog.vala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        custom_bin.add (passkey_label);
+        custom_bin.show_all ();
+    }
+
+    public PairDialog.with_pin_code (ObjectPath device, string pincode) {
+        Object (
+            buttons: Gtk.ButtonsType.CANCEL,
+            image_icon: new ThemedIcon ("dialog-information"),
+            primary_text: _("Confirm Bluetooth PIN"),
+            secondary_text: _("Make sure the code displayed on \"%s\" matches the one below.").printf ("Helix-Sama")
```
Comment (cassidyjames) [inline src/PairDialog.vala]: Typographically-correct quotes:

```suggestion
            secondary_text: _("Make sure the code displayed on “%s” matches the one below.").printf ("Helix-Sama")
```

**Comment by cassidyjames** [comment]
Would something like this work? I guess we can be even smarter about that second text, given we should know the device type. But I'm mainly wondering about the code being inline and bold.

image

**Comment by cassidyjames** [comment]
I was also going to suggest badging the Bluetooth icon with the device type icon to make it more like other agents, but realize that would imply dropping to a Gtk.Dialog or implementing https://github.com/elementary/granite/issues/136

**Comment by danrabbit** [comment]
@cassidyjames Yeah I'd really like to get the device icon in here, so I'm interested in implementing badging in Granite.

I'd like to keep the code on its own line and much larger so that it's easy to see, especially if you're paring from a different device like a phone and you're not necessarily sitting at your desk

**Comment by peteruithoven** [comment]
Pairing, connecting and re-connecting seems to still work for:
- Bose QuietControl 30 (headphones)
- Anker SoundBunds Slim (headphones)
- UE Boom 2 (speaker)
- Logitech M720 Triathlon (mouse
- Logi R500 (Pointer)

**Comment by danrabbit** [comment]
Merging after verbal approval in Meet

**Pull request merged**